sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -826,13 +844,24 @@ struct _pi_program : _pi_object {
 
 struct _pi_kernel : _pi_object {
   _pi_kernel(ze_kernel_handle_t Kernel, pi_program Program)
-      : ZeKernel{Kernel}, Program{Program} {}
+      : ZeKernel{Kernel}, Program{Program}, MemAllocs{}, UsersCount{0} {}
 
   // Level Zero function handle.
   ze_kernel_handle_t ZeKernel;
 
   // Keep the program of the kernel.
   pi_program Program;
+
+  // If kernel has indirect access we need to make a snapshot of all existing
+  // memory allocations to defer deletion of these memory allocations to the
+  // moment when kernel execution has finished.
+  std::list<MemAllocRecord *> MemAllocs;
+
+  // Track the number of events associated with this kernel.
+  // When this value is zero, it means that kernel is not submitted for an
+  // execution. It allows to release memory allocations referenced by this
+  // kernel.
+  std::atomic<pi_uint32> UsersCount;
```

Comment:
Please clarify relation to the existing `Kernel->RefCount`.

I assume you are adding this to be able to release memory sooner than the kernel would be destroyed via regular reference-counting, right? If yes, then can you justify that this is really needed

Reply:
Yes, you are right. Kernels are cached in the context by SYCL RT and they are release only during context object destruction, this is too late.
Regular RefCount is not usable for this because user/SYCL RT can retain kernel object any number of times. And that's why there is no value of RefCount which can mean zero uses (not submitted for an execution).

Reply:
Please clarify this in source comment. Also maybe name it to be more clear, e.g. SubmissionsCount?

Reply:
Fixed and changed the name.